src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1394,91 +1394,66 @@ RewriteResponse TheoryStringsRewriter::preRewrite(TNode node) {
 Node TheoryStringsRewriter::rewriteContains( Node node ) {
   if( node[0] == node[1] ){
     return NodeManager::currentNM()->mkConst( true );
-  }else if( node[0].isConst() && node[1].isConst() ){
+  }else if( node[0].isConst() ){
     CVC4::String s = node[0].getConst<String>();
-    CVC4::String t = node[1].getConst<String>();
-    if( s.find(t) != std::string::npos ){
-      return NodeManager::currentNM()->mkConst( true );
+    if( node[1].isConst() ){
+      CVC4::String t = node[1].getConst<String>();
+      return NodeManager::currentNM()->mkConst( s.find(t)!=std::string::npos );
     }else{
-      return NodeManager::currentNM()->mkConst( false );
-    }
-  }else if( node[0].getKind()==kind::STRING_CONCAT ){
-    //component-wise containment
-    unsigned n1 = node[0].getNumChildren();
-    std::vector< Node > nc1;
-    getConcat( node[1], nc1 );
-    unsigned n2 = nc1.size();
-    if( n1>n2 ){
-      unsigned diff = n1-n2;
-      for(unsigned i=0; i<diff; i++) {
-        if( node[0][i]==nc1[0] ){
-          bool flag = true;
-          for(unsigned j=1; j<n2; j++) {
-            if( node[0][i+j]!=nc1[j] ){
-              flag = false;
-              break;
-            }
-          }
-          if(flag) {
-            return NodeManager::currentNM()->mkConst( true );
-          }
+      if( s.size()==0 ){
+        return NodeManager::currentNM()->mkNode( kind::EQUAL, node[0], node[1] );
+      }else if( node[1].getKind()==kind::STRING_CONCAT ){
+        int firstc, lastc;
+        if( !canConstantContainConcat( node[0], node[1], firstc, lastc ) ){
+          return NodeManager::currentNM()->mkConst( false );
         }
       }
     }
+  }
+  std::vector< Node > nc1;
+  getConcat( node[0], nc1 );
+  std::vector< Node > nc2;
+  getConcat( node[1], nc2 );
+  std::vector< Node > nr;
+  
+  //component-wise containment
+  if( componentContains( nc1, nc2, nr )!=-1 ){
+    return NodeManager::currentNM()->mkConst( true );
+  }
+  
+  //strip endpoints
+  std::vector< Node > nb;
+  std::vector< Node > ne;
+  if( stripConstantEndpoints( nc1, nc2, nb, ne ) ){
+    return NodeManager::currentNM()->mkNode( kind::STRING_STRCTN, mkConcat( kind::STRING_CONCAT, nc1 ), node[1] );
+  }
+  Trace("strings-rewrite-debug2") << "No constant endpoints for " << node[0] << " " << node[1] << std::endl;
+
+  //splitting
+  if( node[0].getKind()==kind::STRING_CONCAT ){
     if( node[1].isConst() ){
       CVC4::String t = node[1].getConst<String>();
-      for(unsigned i=0; i<node[0].getNumChildren(); i++){
+      for(unsigned i=1; i<(node[0].getNumChildren()-1); i++){
```

Comment:
Could we add an assert that says that `node[0][0]` and `node[0][node[0].getNumChildren() - 1]` are both not const? My understanding is that this is the case because of `stripConstantEndpoints` but it would be good to have the asserts here and maybe a short comment because it is not very obvious.

Reply:
Actually I think there are cases where this assertion does not hold.

Take str.contains( str.++( "a", x, "c", y, "b" ), "ab" ).

In this case, we cannot strip the first component, nor the last component (since x may start with "b", or y may end with "a").

Reply:
Regardless, I'll add a comment to explain why we only consider the inner children.